cdap-unit-test/src/test/java/co/cask/cdap/service/FileUploadServiceTestRun.java

```diff
@@ -0,0 +1,120 @@
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ *
+ */
+public class FileUploadServiceTestRun extends TestFrameworkTestBase {
+
+  @Test
+  public void testFileUploadService() throws Exception {
+    ApplicationManager appManager = deployApplication(FileUploadApp.class);
+
+    // Start the service
+    ServiceManager serviceManager = appManager.getServiceManager(FileUploadApp.SERVICE_NAME).start();
+    try {
+      // Upload URL is "base/upload/pfs/[partition_value], which the partition value is a long
+      URI serviceURI = serviceManager.getServiceURL(10, TimeUnit.SECONDS).toURI();
+
+      // Upload with wrong MD5, should get 400.
+      byte[] content = Strings.repeat("0123456789 ", 100).getBytes(Charsets.UTF_8);
+      Assert.assertEquals(HttpURLConnection.HTTP_BAD_REQUEST,
+                          upload(serviceURI.resolve("upload/" + FileUploadApp.PFS_NAME + "/1").toURL(), content,
+                                 "123", 30));
+
+      // Upload with right MD5, should get 200
+      Assert.assertEquals(HttpURLConnection.HTTP_OK,
+                          upload(serviceURI.resolve("txUpload/" + FileUploadApp.PFS_NAME + "/1").toURL(), content,
+                                 BaseEncoding.base64().encode(Hashing.md5().hashBytes(content).asBytes()), 20));
+
+      // Inspect the partitioned file set and verify the content
+      PartitionedFileSet pfs = (PartitionedFileSet) getDataset(FileUploadApp.PFS_NAME).get();
+      PartitionDetail partition = pfs.getPartition(PartitionKey.builder().addLongField("time", 1).build());
+
+      Assert.assertNotNull(partition);
+
+      // There should be one file under the partition directory
+      List<Location> locations = partition.getLocation().list();
+      Assert.assertEquals(1, locations.size());
+      Assert.assertArrayEquals(content, ByteStreams.toByteArray(Locations.newInputSupplier(locations.get(0))));
+
+      // Verify the tracking table of chunks sizes
+      KeyValueTable trackingTable = (KeyValueTable) getDataset(FileUploadApp.KV_TABLE_NAME).get();
+      CloseableIterator<KeyValue<byte[], byte[]>> iter = trackingTable.scan(null, null);
+      // Sum up all chunks sizes as being tracked by the tracking table.
+      long sum = 0;
+      int iterSize = 0;
+      while (iter.hasNext()) {
+        KeyValue<byte[], byte[]> kv = iter.next();
+        sum += Bytes.toInt(kv.getKey()) * Bytes.toLong(kv.getValue());
+        iterSize++;
+      }
+      // The iterator should have size >= 2, since we uses different chunk size for two different upload
+      Assert.assertTrue(iterSize >= 2);
+
+      // The sum of all chunks sizes should be the same as the
+      // content size * 2 (since we have one failure, one success upload)
+      Assert.assertEquals(content.length * 2, sum);
+
+    } finally {
```

Comment:
stopping programs isn't necessary in user test cases anymore.
All programs will get stopped in: TestBase#afterTest